### PR TITLE
sql: turn ranges table into a view

### PR DESCRIPTION
**Pull request by changangela**
Related to #30683.

We want to build `crdb_internal.ranges` table as a view over
`crdb_internal.ranges_no_leases` in order to leverage the existing opt rules in
pushing down filters before the expensive computation on the `leaseholder`
column.

Release note: None

### CODE REVIEW

File: pkg/sql/virtual_schema.go [hunk 1]
```diff
@@ -71,6 +76,95 @@ type virtualSchemaTable struct {
...
+	if err != nil {
+		return sqlbase.TableDescriptor{}, err
+	}
+
+	create := stmt.(*tree.CreateView)
+
+	return MakeViewTableDesc(
+		create,
+		v.getResultColumns(parentDesc),
+		keys.DescriptorTableID,
```
Comment (changangela) [inline pkg/sql/virtual_schema.go]: For some reason, using `0` as `parentID` and `keys.VirtualDescriptorID` as `ID` causes logictest to fail on the query `SELECT * FROM crdb_internal.ranges` (but this query works when using `./cockroach sql`).

**Review by RaduBerinde — COMMENTED** [review]
Thanks for reworking things with this approach!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-:-LNl5zPY91nhos6oMneC:b-l8l23j)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/crdb_internal.go, line 1551 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNl0T_WDEd6Sv2o9d2L:-LNl0T_WDEd6Sv2o9d2M:b-aea531) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/crdb_internal.go#L1551)):*
> ```Go
> 	end_pretty,
> 	database,
> 	"table",
> ```

do we need these quotes?
___
*[pkg/sql/crdb_internal.go, line 1557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNl618t8ELguPn-kpgq:-LNl618t8ELguPn-kpgr:b-b27ibl) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/crdb_internal.go#L1557)):*
> ```Go
> FROM crdb_internal.ranges_no_leases
> `,
> 	parent: "ranges_no_leases",
> ```

I'm not crazy about the "parent" approach. We only use this to deduce the columns of the view, correct? It seems unnecessarily restrictive - a view could in principle depend on multiple tables; or it could add the new column in a different position. I'd just write down all the result columns here
```
  sqlbase.ResultColumns{
    {Name: "range_id", Typ: types.Int},
    {Name: "start_key", Typ: types.Bytes},
    ..
  }
```
___
*[pkg/sql/virtual_schema.go, line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNl10vM7oiPUfgVNECw:-LNl10vM7oiPUfgVNECx:b-vzeu32) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/virtual_schema.go#L83)):*
> ```Go
> 	schema string
> 
> 	// parent represents the table that the virtualSchemaView is augmenting
> ```

[nit] add `.` at the end
___
*[pkg/sql/virtual_schema.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNl0fmf2j0u9TNRyb-o:-LNl0fmf2j0u9TNRyb-p:b-2m0fb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/virtual_schema.go#L90)):*
> ```Go
> }
> 
> func (t virtualSchemaTable) getSchema() string {
> ```

[nit] add a method comment `// getSchema is part of the virtualSchemaDef interface.`
___
*[pkg/sql/virtual_schema.go, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNl0x6vBWrPPnbeymuk:-LNl0x6vBWrPPnbeymul:b-zfpzi3) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/virtual_schema.go#L94)):*
> ```Go
> }
> 
> func (t virtualSchemaView) getSchema() string {
> ```

[nit] we shouldn't interleave methods for one type with those for another
___
*[pkg/sql/virtual_schema.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNkz3t7DCwC_z2xXzng-r1-146:-LNl2Dpq1hbzfEmPUNIw:brpol0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/virtual_schema.go#L146)):*
<details><summary><i>Previously, changangela (Angela Chang) wrote…</i></summary><blockquote>

For some reason, using `0` as `parentID` and `keys.VirtualDescriptorID` as `ID` causes logictest to fail on the query `SELECT * FROM crdb_internal.ranges`, `getPlanInfo` would be called on a  `virtualSchemaView` which produces a panic. However, this same query worked fine when using `./cockroach sql`. I'm not too sure what the best way to fix this is.
</blockquote></details>

Passing `keys.SystemDatabaseID` as the parentID would make sense to me, but we're passing it as `id`. Using `DescriptorTableID` for either seems wrong.

What was the error?
___
*[pkg/sql/virtual_schema.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNl1gbi44Hg257devjl:-LNl1gbi44Hg257devjm:b-1iu3s3) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/virtual_schema.go#L150)):*
> ```Go
> 		hlc.Timestamp{},
> 		publicSelectPrivileges,
> 		nil,
> ```

[nit] add comments `// semaCtx`, `// evalCtx` so it's clear what these fields are
___
*[pkg/sql/virtual_schema.go, line 308 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNl4aibDTtGJ0BGJoMr:-LNl4aibDTtGJ0BGJoMs:b9090ag) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/virtual_schema.go#L308)):*
> ```Go
> 			var err error
> 
> 			switch def.(type) {
> ```

you can do `switch v := def.(type) {` and then use `v` as `virtualSchemaTable` or `virtualSchemaView` in each case

**Review by changangela — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/crdb_internal.go, line 1551 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNl0T_WDEd6Sv2o9d2L:-LNl9gWj5Sg7SxyblqYR:b-1o8ndn) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/crdb_internal.go#L1551)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

do we need these quotes?
</blockquote></details>

I think so, removing them results in the following error: `: syntax error at or near "table"`
___
*[pkg/sql/virtual_schema.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNkz3t7DCwC_z2xXzng-r1-146:-LNlBrsj0HghFpo7bTYV:bfhyy8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/virtual_schema.go#L146)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Passing `keys.SystemDatabaseID` as the parentID would make sense to me, but we're passing it as `id`. Using `DescriptorTableID` for either seems wrong.

What was the error?
</blockquote></details>

`interface conversion: sql.virtualSchemaDef is sql.virtualSchemaView, not sql.virtualSchemaTable` - the `getPlanInfo` should not be called on a `virtualSchemaView` in theory.

**Review by knz — COMMENTED** [review]
I like the direction this is going. See below for suggestions to unblock.

> Reviewed 4 of 6 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-:-LNoBT-c5NYvUJGNVRvf:b-r7suio)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/crdb_internal.go, line 1551 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNl0T_WDEd6Sv2o9d2L:-LNo9e5i2_tMqgIGNtKz:b69529m) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/crdb_internal.go#L1551)):*
<details><summary><i>Previously, changangela (Angela Chang) wrote…</i></summary><blockquote>

I think so, removing them results in the following error: `: syntax error at or near "table"`
</blockquote></details>

Please read the description of PR #27098 in particular the sections "Motivation", "Problem Statement" and "Methodology"

Then please rename the 3 columns here accordingly.

(I apologize this information is currently only available in this PR. We ought to have a common doc available to everyone who starts working in SQL)
___
*[pkg/sql/virtual_schema.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNkz3t7DCwC_z2xXzng-r1-146:-LNoADjd16wyWTrMjjX3:bt9nwmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/38de9d79465d5844bd6802b89746ba31a5ba061d/pkg/sql/virtual_schema.go#L146)):*
<details><summary><i>Previously, changangela (Angela Chang) wrote…</i></summary><blockquote>

`interface conversion: sql.virtualSchemaDef is sql.virtualSchemaView, not sql.virtualSchemaTable` - the `getPlanInfo` should not be called on a `virtualSchemaView` in theory.
</blockquote></details>

The parent ID should be 0 otherwise other things will become confused.

I do not understand why you think it's not reasonable to have `getPlanInfo` implemented for `virtualSchemaView`.
___
*[pkg/sql/virtual_schema.go, line 218 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNoAzDk1t1nu3KTzo0m:-LNoAzDk1t1nu3KTzo0n:b-anbxgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abefa8b13361403a60df79cdadbd9531e418bda/pkg/sql/virtual_schema.go#L218)):*
> ```Go
> 
> 	constructor := func(ctx context.Context, p *planner, dbName string) (planNode, error) {
> 		tableDesc := e.tableDef.(virtualSchemaTable)
> ```

Here is your problem. If `getPlanInfo` is called on a virtual view definition, it must recurse via `(*planner).getViewPlan` instead of returning a `constructor`.

Of course in the context  of the optbuilder this cannot use the heuristic planner so you need a callback as argument that "does the right thing" -- the call back will be either an optbuilder closure or a heuristic planner closure. @RaduBerinde thoughts?

**Review by changangela — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/virtual_schema.go, line 218 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNoAzDk1t1nu3KTzo0m:-LNpMbUn3Wbs5aa8fZaJ:bej6ff5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abefa8b13361403a60df79cdadbd9531e418bda/pkg/sql/virtual_schema.go#L218)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Here is your problem. If `getPlanInfo` is called on a virtual view definition, it must recurse via `(*planner).getViewPlan` instead of returning a `constructor`.

Of course in the context  of the optbuilder this cannot use the heuristic planner so you need a callback as argument that "does the right thing" -- the call back will be either an optbuilder closure or a heuristic planner closure. @RaduBerinde thoughts?
</blockquote></details>

Thanks @knz! I added the recursive call to `(*planner).getViewPlan` and the broken logictest passes. I'm a bit confused about what the callback is for?

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/virtual_schema.go, line 218 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNoAzDk1t1nu3KTzo0m:-LNpTRJeEAcxPme8JfWS:bgyaotk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abefa8b13361403a60df79cdadbd9531e418bda/pkg/sql/virtual_schema.go#L218)):*
<details><summary><i>Previously, changangela (Angela Chang) wrote…</i></summary><blockquote>

Thanks @knz! I added the recursive call to `(*planner).getViewPlan` and the broken logictest passes. I'm a bit confused about what the callback is for?
</blockquote></details>

`getViewPlan` is code from heuristic planner, it won't use the optimizer. But that's probably good enough for now, the heuristic planner should be able to do the filter pushdown we want. I'd leave a `TODO` where we call `getViewPlan` mentioning this though.

**Review by knz — COMMENTED** [review]
> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/virtual_schema.go, line 218 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LNoAzDk1t1nu3KTzo0m:-LNpVvW31qogaUXuxL94:b-lrzvs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abefa8b13361403a60df79cdadbd9531e418bda/pkg/sql/virtual_schema.go#L218)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`getViewPlan` is code from heuristic planner, it won't use the optimizer. But that's probably good enough for now, the heuristic planner should be able to do the filter pushdown we want. I'd leave a `TODO` where we call `getViewPlan` mentioning this though.
</blockquote></details>

@raduberinde if the optimizer kicks in and there is a planNode tree built by `getViewPlan`, this plan will not get processed by `expandPlan` (IIRC, `expandPlan` is not used when the opt code was successful). So filter propagation won't occur. Are you sure there's no way to inject the optbuilder here?

**Review by RaduBerinde — COMMENTED** [review]
@changangela can you run `explain (verbose) select * from crdb_internal.ranges where range_id = 10;` and `explain (opt)` for the same and post the output?

This is the `explain (opt)` I get if I just create the view locally; this is what it should look like: https://gist.github.com/RaduBerinde/bdab4dc719348a9258a9f302d1d032ce

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-:-LNpheBc3u5Qrgre57XU:b-iuvivl)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by changangela — COMMENTED** [review]
`explain(verbose)` output:
https://gist.github.com/changangela/434e2bec2d22f062284bebeae1040d75

`explain(opt)` output (looks as expected):
https://gist.github.com/changangela/f8a6ea0b9d028d641cfe80d054c7912c

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-:-LNpj73EDkztdZBUQrwa:btx3k9i)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by knz** [comment]
Radu and I were talking about this.

- as your EXPLAIN(OPT) output confirms, the optbuilder (and generally the opt code) is handling all this properly already. So my idea to use a callback is not necessary.

- the problem with the heuristic planner is in a different place. The heuristic planner should test whether a descriptor refers to a view **before** it tests whether it's a virtual table. Currently this is done incorrectly. The code that is incorrect is the case for `*tree.NormalizableTableName` in `getDataSource`. 

Radu suggests to modify the method `IsVirtualTable()` on descriptors to also check `!IsView()`. I am not yet 100% sold on this -- you'd need first to check all the callers to verify this fix is appropriate. If it is indeed appropriate go that way; otherwise, simply add the `&& !IsView()` to `getDataSource`.

**Review by RaduBerinde — COMMENTED** [review]
Thanks, the output looks good!

Yes, we should fix `getDataSource` to not go down the virtual path for this view. If we fix that we should not need to implement `getPlanInfo` for the view (the "regular" view code would handle it).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-:-LNpkDqq0TSYNjpdzApE:buzc7u8)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by changangela — COMMENTED** [review]
Adding `!isView()` inside `IsVirtualTable()` yields the following error:
```
 failed to initialize
CREATE VIEW crdb_internal.ranges AS SELECT
	range_id,
	start_key,
	start_pretty,
	end_key,
	end_pretty,
	database_name,
	table_name,
	index_name,
	replicas,
	crdb_internal.lease_holder(start_key) AS lease_holder
FROM crdb_internal.ranges_no_leases
: invalid parent ID 0
```

Are there any valid parent ID's I can give to this view? Otherwise, I'll look into adding `IsView()` to `getDataSource`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-:-LNplhkK2lmodTxdMPsN:bax9xtl)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by changangela — COMMENTED** [review]
Edit: I added the `&& !isView()` check into `getDataSource`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-:-LNpnW7l4yIfleVITAmu:b-g982px)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by RaduBerinde — COMMENTED** [review]
:lgtm:

Thanks for your patience on this!

I would grep for all instances of `crdb_internal.ranges` in the codebase and see if anything needs updating (because of the new column names), like `TestClosedTimestampCanServe`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-:-LNprY2l7Q9ntYXDZmQj:b-7kyh7e)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30846)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/clearrange.go, line 102 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/30846#-LOLGJEu3R_WqKtDqD8C:-LOLGJEu3R_WqKtDqD8D:b2kozq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/13ee419292a1cd70b0f8f490dac3053a2d7c531a/pkg/cmd/roachtest/clearrange.go#L102)):*
> ```Go
> 		if true {
> 			if err := conn.QueryRow(
> 				`SELECT to_hex(start_key) FROM crdb_internal.ranges WHERE database_name = 'bank' AND table_name = 'bank' ORDER BY start_key ASC LIMIT 1`,
> ```

FYI, this broke the `clearrange` test when run on `release-2.1`. Somewhat surprisingly, we use the `roachtest` from `master` when running against `release-2.1` (and `release-2.0`). I think you can use `test.IsBuildVersion()` to change the query depending on the version of Cockroach we're running against.

**Comment by changangela** [comment]
@petermattis should we change the column name on release-2.1 instead of switching on the build version?

**Comment by petermattis** [comment]
@changangela That works for me.

**Pull request merged**